pkg/filepromoter/gcs.go

```diff
@@ -46,11 +52,57 @@ func (s *gcsSyncFilestore) OpenReader(
 	return s.client.Bucket(s.bucket).Object(absolutePath).NewReader(ctx)
 }
 
-// UploadFile uploads a local file to the specified destination
+// UploadFileRaw uploads a local file to the specified destination.
+// It will attempt to use transparent compression for faster/cheaper downloads.
 func (s *gcsSyncFilestore) UploadFile(
 	ctx context.Context,
 	dest string,
 	localFile string) error {
+
+	sha256, err := computeSHA256ForFile(localFile)
+	if err != nil {
+		return err
+	}
+
+	stat, err := os.Stat(localFile)
+	if err != nil {
+		return fmt.Errorf("error getting stat for %q: %v", localFile, err)
+	}
+
+	metadata := map[string]string{}
+
+	tmpfile, err := maybeGzip(localFile)
+	if err != nil {
+		return fmt.Errorf("error compressing file: %v", err)
+	}
+	if tmpfile == "" {
+		// Not worth compressing
+
+		metadata[MetadataKeySHA256] = sha256
+
+		return s.uploadFileRaw(ctx, dest, localFile, "", "", metadata)
+	}
+
+	defer loggedRemove(tmpfile)
+
+	contentEncoding := "gzip"
+	contentType := "application/octet-stream"
```

Comment:
Ah OK, so we just compress to gzip when we upload it, but on the GCS server it will (of course) de-compress it back. Hence the phrase "transparent compression".

I like to state the obvious to make sure that I understand what's going on... :)